derive.h

```diff
@@ -0,0 +1,311 @@
+							)
+						)
+					);
+					return multiply(be1, lhs->derivative());
+				}
+			default: assert(0); 
+		}
+	}
+
+protected:
+
+	virtual optional<char> opt_var() const override { 
+		if (!lhs->opt_var()) {
+			return rhs->opt_var();
+		}
+
+		if (!rhs->opt_var()) {
+			return lhs->opt_var();
+		}
+	
+		assert(lhs->opt_var() == rhs->opt_var());
+		return lhs->opt_var();
+	}
+	shared_ptr<Expression> lhs;
+	shared_ptr<Expression> rhs;
+	Operation operation;
+};
+
+shared_ptr<Expression> make_shared_expression(CoefExpression* ptr) {
+	return shared_ptr<CoefExpression>(ptr);
+}
+
+shared_ptr<Expression> make_shared_expression(BinaryExpression* ptr) {
+	return shared_ptr<BinaryExpression>(ptr);
+}
+
+shared_ptr<Expression> add(const shared_ptr<Expression>& lhs, const shared_ptr<Expression>& rhs) {
+	auto coef_left = std::dynamic_pointer_cast<CoefExpression>(lhs);
+	auto coef_right = std::dynamic_pointer_cast<CoefExpression>(rhs);
+	if (coef_left && coef_right) {
+		return make_shared_expression(
+			new CoefExpression{coef_left->coef() + coef_right->coef()}
+		); 
+	}
+
+	if (lhs->is_zero()) {
+		return rhs;
+	}
+	
+	if (rhs->is_zero()) {
+		return lhs;
+	}
+	
+	return make_shared_expression(
+		new BinaryExpression{lhs, rhs, plus}
+	);
+}
+
+shared_ptr<Expression> subtract(const shared_ptr<Expression>& lhs, const shared_ptr<Expression>& rhs) {
+	auto coef_left = std::dynamic_pointer_cast<CoefExpression>(lhs);
+	auto coef_right = std::dynamic_pointer_cast<CoefExpression>(rhs);
+	if (coef_left && coef_right) {
+		return make_shared_expression(
+			new CoefExpression{coef_left->coef() - coef_right->coef()}
+		); 
+	}
+
+	if (rhs->is_zero()) {
+		return lhs;
+	}
+	
+	return make_shared_expression(
+		new BinaryExpression{lhs, rhs, minus}
+	);
+}
+
+shared_ptr<Expression> multiply(const shared_ptr<Expression>& lhs, const shared_ptr<Expression>& rhs) {
+	auto coef_left = std::dynamic_pointer_cast<CoefExpression>(lhs);
+	auto coef_right = std::dynamic_pointer_cast<CoefExpression>(rhs);
+	if (coef_left && coef_right) {
+		return make_shared_expression(
+			new CoefExpression{coef_left->coef() * coef_right->coef()}
+		); 
+	}
+	if (
+		rhs->is_zero() || lhs->is_zero()
+	) {
+		return make_shared_expression(
+			new CoefExpression{0.}
+		); 
+	}
+
+	return make_shared_expression(
+		new BinaryExpression{lhs, rhs, mult}
+	);
+}
+
+shared_ptr<Expression> pow(const shared_ptr<Expression>& lhs, const shared_ptr<Expression>& rhs) {
+	return make_shared_expression(
+		new BinaryExpression{lhs, rhs, power}
+	);
+}
+
+} // End of namespace detail
+
+class Func {
+public:
+	Func(double coef) :
+		m_expr(std::dynamic_pointer_cast<detail::Expression>(
+			shared_ptr<detail::CoefExpression>(new detail::CoefExpression{coef})
+		))
+	{ }	
+
+	Func(char x):
+		m_expr(std::dynamic_pointer_cast<detail::Expression>(
+			shared_ptr<detail::XExpression>(new detail::XExpression{x})
+		))
+	{ }	
+
+	Func(Func&& func) : 
+		m_expr(func.m_expr) {
```

Comment:
```
m_expr(std::move(func.m_expr))
```
also then there's no need to do func.m_expr = nullptr

Reply:
Right! Fixed.